presto-hive/src/main/java/com/facebook/presto/hive/SqlStandardAccessControl.java

```diff
@@ -189,24 +192,40 @@ public void checkCanGrantTablePrivilege(Identity identity, Privilege privilege,
         }
 
         HivePrivilege hivePrivilege = toHivePrivilege(privilege);
-        if (hivePrivilege == null || !metastore.hasPrivilegeWithGrantOptionOnTable(identity.getUser(), tableName.getSchemaName(), tableName.getTableName(), hivePrivilege)) {
+        if (hivePrivilege == null || !getGrantOptionForPrivilege(identity, privilege, tableName)) {
             denyGrantTablePrivilege(privilege.name(), tableName.toString());
         }
     }
 
-    private boolean checkDatabasePermission(Identity identity, String schemaName, HivePrivilege... requiredPrivileges)
+    private boolean checkDatabasePermission(Identity identity, String schemaName, HivePrivilegeInfo.HivePrivilege... requiredPrivileges)
     {
-        Set<HivePrivilege> privilegeSet = metastore.getDatabasePrivileges(identity.getUser(), schemaName);
+        Set<HivePrivilegeInfo.HivePrivilege> privilegeSet = metastore.getDatabasePrivileges(identity.getUser(), schemaName).stream()
+                .map(hivePrivilegeInfo -> hivePrivilegeInfo.getHivePrivilege())
+                .collect(Collectors.toSet());
+
         return privilegeSet.containsAll(ImmutableSet.copyOf(requiredPrivileges));
     }
 
-    private boolean checkTablePermission(Identity identity, SchemaTableName tableName, HivePrivilege... requiredPrivileges)
+    private boolean getGrantOptionForPrivilege(Identity identity, Privilege privilege, SchemaTableName tableName)
+    {
+        Set<HivePrivilegeInfo> privilegeSet = metastore.getTablePrivileges(identity.getUser(), tableName.getSchemaName(), tableName.getTableName());
+        for (HivePrivilegeInfo privilegeInfo : privilegeSet) {
```

Comment:
done in https://github.com/Teradata/presto/pull/166/commits/05c8f1b17d9b80fe8a95a15146eb66058b46f40e